Cesium.CodeGen/Ir/Expressions/IdentifierExpression.cs

```diff
@@ -32,27 +35,44 @@ public IdentifierExpression(Ast.IdentifierExpression expression)
     public ILValue Resolve(IDeclarationScope scope)
     {
         scope.Variables.TryGetValue(Identifier, out var var);
+        scope.Functions.TryGetValue(Identifier, out FunctionInfo? fun);
         var par = scope.GetParameter(Identifier);
         scope.Context.AssemblyContext.GlobalFields.TryGetValue(Identifier, out var globalType);
-        switch (var, par)
+
+        if (var is not null && par is not null)
+            throw new NotSupportedException($"Variable {Identifier} is both available as a local and as a function parameter.");
+
+        if (var is not null && fun is not null)
+            throw new NotSupportedException($"Variable {Identifier} is both available as a local and as a function name.");
+
+        if (fun is not null && par is not null)
+            throw new NotSupportedException($"Variable {Identifier} is both available as a function name and as a function parameter.");
+
+        if (var is not null)
+        {
+            var variableDefinition = scope.ResolveVariable(Identifier);
+            return new LValueLocalVariable(variableDefinition);
+        }
+
+        if (par is not null)
         {
-            case (null, null):
-                if (globalType != null)
-                {
-                    var flobalField = scope.Context.AssemblyContext.ResolveGlobalField(Identifier, scope.Context);
-                    return new LValueGlobalVariable(flobalField);
-                }
-                throw new NotSupportedException($"Cannot find variable {Identifier}.");
-            case ({ }, null):
-                {
-                    var variableDefinition = scope.ResolveVariable(Identifier);
-                    return new LValueLocalVariable(variableDefinition);
-                }
-            case (null, { }):
-                return new LValueParameter(par);
-            case ({ }, { }):
-                throw new NotSupportedException(
-                    $"Variable {Identifier} is both available as a local and as a function parameter.");
+            return new LValueParameter(par);
         }
+        
+        if (fun is not null)
+        {
+            var functionType = new FunctionType(fun.Parameters, fun.ReturnType);
+            var variableDefinition = new VariableDefinition(functionType.ResolvePointer(scope.Context));
+            return new LValueLocalVariable(variableDefinition);
```

Comment:
Free function shouldn't be represented by `LValueLocalVariable`; creating a `VariableDefinition` in this point is wrong.

Moreover, after I've taken a look at the code using `IdentifierExpression`, I now think that maybe a better abstraction should be used.

Let's create a new interface, `IValue` (without the `EmitSetValue` method), and inherit `ILValue` from it.

`IdentifierExpression::Resolve` would then return an `IValue` and not an `ILValue`. And for a function, let's create a new implementation of `IValue`.

Reply:
Thx. I was thinking about this. That is what I mean in my `shouldn't be, is it?`. Will try to do it.